### PR TITLE
[Translation] added option json_options to JsonFileDumper.

### PR DESCRIPTION
**Pull request by gepo**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | no |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

It's passed to json_encode function second argument.

### CODE REVIEW

File: src/Symfony/Component/Translation/Dumper/JsonFileDumper.php [hunk 1]
```diff
@@ -23,9 +23,23 @@ class JsonFileDumper extends FileDumper
     /**
      * {@inheritdoc}
```
Comment (ogizanagi) [inline src/Symfony/Component/Translation/Dumper/JsonFileDumper.php]: Wrong docblock. This method isn't inherited from anything.

File: src/Symfony/Component/Translation/Dumper/JsonFileDumper.php [hunk 2]
```diff
@@ -23,9 +23,23 @@ class JsonFileDumper extends FileDumper
...
      */
+    public function formatCatalogue(MessageCatalogue $messages, $domain, array $options = array())
```
Comment (ogizanagi) [inline src/Symfony/Component/Translation/Dumper/JsonFileDumper.php]: How are set those options ? I mean, the method is never called with the options parameter ?

File: src/Symfony/Component/Translation/Dumper/JsonFileDumper.php (hunk 1 above)
Comment (gepo) [inline src/Symfony/Component/Translation/Dumper/JsonFileDumper.php]: @ogizanagi 
this code inherit my previous, accepted PR:
https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Translation/Dumper/FileDumper.php

Comment (gepo) [inline src/Symfony/Component/Translation/Dumper/JsonFileDumper.php]: it's 2.8 branch

Comment (aitboudad) [inline src/Symfony/Component/Translation/Dumper/JsonFileDumper.php]: see https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Translation/Dumper/FileDumper.php#L100

Comment (ogizanagi) [inline src/Symfony/Component/Translation/Dumper/JsonFileDumper.php]: Oh, sorry. Know I understand. Please forget about my next comment.

File: src/Symfony/Component/Translation/Dumper/JsonFileDumper.php (hunk 2 above)
Comment (aitboudad) [inline src/Symfony/Component/Translation/Dumper/JsonFileDumper.php]: it should be `protected` no ?

**Comment by aitboudad** [comment]
can we found another name rather than `json_options` ?
The main issue here if we add this options it should be fully compatible with `JsonFileLoader` else I'm :-1:

File: src/Symfony/Component/Translation/Dumper/JsonFileDumper.php (hunk 2 above)
Comment (gepo) [inline src/Symfony/Component/Translation/Dumper/JsonFileDumper.php]: @aitboudad yes, you are right, change it to protected
@ogizanagi yep, change it to JSON_HEX_QUOT, it's from 5.3.

**Comment by gepo** [comment]
@aitboudad I will think about better name, just 'options' are even worse.

Other alternative is to add several options - one per JSON_* constants and hide from user, that this class use json_encode internally. But it would be hard to maintain such code.

**Comment by aitboudad** [comment]
`encode_options` or `bitmasks` (an array of bitmask options) ?

**Comment by dosten** [comment]
I prefer "json_encoding" for consistency. see https://github.com/symfony/symfony/blob/2.8/src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/JsonDescriptor.php#L173

File: src/Symfony/Component/Translation/Dumper/JsonFileDumper.php [hunk 3]
```diff
@@ -23,9 +23,23 @@ class JsonFileDumper extends FileDumper
     /**
      * {@inheritdoc}
      */
+    protected function formatCatalogue(MessageCatalogue $messages, $domain, array $options = array())
+    {
+        $jsonOptions = isset($options['json_options']) ? (int) $options['json_options'] : 0;
```
Comment (dosten) [inline src/Symfony/Component/Translation/Dumper/JsonFileDumper.php]: I prefer `$flags` instead of `$jsonOptions`

**Comment by aitboudad** [comment]
@dosten :+1:

**Comment by gepo** [comment]
@dosten :+1: apply

File: src/Symfony/Component/Translation/Dumper/JsonFileDumper.php [hunk 4]
```diff
@@ -23,9 +23,23 @@ class JsonFileDumper extends FileDumper
     /**
      * {@inheritdoc}
      */
+    protected function formatCatalogue(MessageCatalogue $messages, $domain, array $options = array())
+    {
+        $flags = isset($options['flags']) ? (int) $options['flags'] : 0;
```
Comment (dosten) [inline src/Symfony/Component/Translation/Dumper/JsonFileDumper.php]: This is wrong, the variable name must be `$flags` but the option name must be "json_encoding"

**Comment by gepo** [comment]
@dosten done

**Comment by dosten** [comment]
:+1:

**Comment by gepo** [comment]
sorry, missed your message. Done

**Comment by dosten** [comment]
:+1:

**Comment by aitboudad** [comment]
ping @symfony/deciders

File: src/Symfony/Component/Translation/Dumper/JsonFileDumper.php [hunk 5]
```diff
@@ -23,9 +23,23 @@ class JsonFileDumper extends FileDumper
     /**
      * {@inheritdoc}
      */
+    protected function formatCatalogue(MessageCatalogue $messages, $domain, array $options = array())
+    {
+        $flags = isset($options['json_encoding']) ? $options['json_encoding'] : 0;
+
+        if (defined('JSON_PRETTY_PRINT')) {
+            $flags |= JSON_PRETTY_PRINT;
```
Comment (stof) [inline src/Symfony/Component/Translation/Dumper/JsonFileDumper.php]: should we really force it all them time, even when encoding options are specified ? this forbids disabling the pretty printing

Comment (aitboudad) [inline src/Symfony/Component/Translation/Dumper/JsonFileDumper.php]: @gepo any change to fix this :)

Comment (gepo) [inline src/Symfony/Component/Translation/Dumper/JsonFileDumper.php]: rewrite this code, do not force pretty print now

Comment (aitboudad) [inline src/Symfony/Component/Translation/Dumper/JsonFileDumper.php]: sound good to me thanks ^^

**Comment by xabbuh** [comment]
Is there a reason to modify the existing test instead of adding an extra one that uses custom options? If I don't miss anything we wouldn't have a test for the default options anymore.

**Comment by aitboudad** [comment]
@gepo Do you have a little time to fix the latest remark :) ?

**Comment by aitboudad** [comment]
Closing in favor of #15756

**Pull request closed without merging**